subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c

```diff
@@ -234,6 +237,87 @@ static int handle_device_config_update(const struct nct_evt *const evt,
 	return err;
 }
 
+static int _log_level;
+
+/* Placeholder until cloud logging added in another PR */
+void nrf_cloud_log_control_set(int log_level)
+{
+	LOG_DBG("Setting nRF Cloud log level = %d", log_level);
+	_log_level = log_level;
+}
+
+/* Placeholder until cloud logging added in another PR */
+int nrf_cloud_log_control_get(void)
+{
+	return _log_level;
+}
+
+static int handle_device_control_update(const struct nct_evt *const evt,
+					bool *const control_found)
+{
+	int err;
+	struct nrf_cloud_evt cloud_evt = {
+		.type = NRF_CLOUD_EVT_RX_DATA_SHADOW
+	};
+	enum nrf_cloud_ctrl_status status = NRF_CLOUD_CTRL_NOT_PRESENT;
+	struct nrf_cloud_ctrl_data ctrl_data;
+
+	if ((evt == NULL) || (control_found == NULL)) {
+		return -EINVAL;
+	}
+
+	if (evt->param.cc == NULL) {
+		return -ENOENT;
+	}
+
+	ctrl_data.alerts_enabled = nrf_cloud_alert_control_get();
+	ctrl_data.log_level = nrf_cloud_log_control_get();
+
+	err = nrf_cloud_decode_control(&evt->param.cc->data, &status, &ctrl_data);
+	if (err) {
+		return (err == -ESRCH) ? 0 : err;
+	}
+
+	*control_found = (status != NRF_CLOUD_CTRL_NOT_PRESENT);
+	if (*control_found) {
+		nrf_cloud_alert_control_set(ctrl_data.alerts_enabled);
+		nrf_cloud_log_control_set(ctrl_data.log_level);
+	}
+
+	/* Acknowledge that shadow delta changes have been made. */
+	if (status == NRF_CLOUD_CTRL_REPLY) {
+		struct nct_cc_data msg = {
+			.opcode = NCT_CC_OPCODE_UPDATE_REQ,
+			.message_id = NCT_MSG_ID_STATE_REPORT,
+		};
+
+		err = nrf_cloud_encode_control_response(&ctrl_data, &msg.data);
+		if (err) {
+			LOG_ERR("nrf_cloud_encode_control_response failed %d", err);
+			return err;
+		}
+
+		if (msg.data.ptr) {
+			err = nct_cc_send(&msg);
+			nrf_cloud_free((void *)msg.data.ptr);
+
+			if (err) {
+				LOG_ERR("nct_cc_send failed %d", err);
+			}
+		}
+	}
+
+	cloud_evt.data = evt->param.cc->data;
+	cloud_evt.topic = evt->param.cc->topic;
+
+	/* Pass the current state since the state is not changing, but we do
+	 * want to let the application get a chance to see this delta too.
+	 */
+	nfsm_set_current_state_and_notify(nfsm_get_current_state(), &cloud_evt);
+
+	return err;
```

Comment:
I apologize that I didn't have the clarity of mind to suggest this when I last reviewed, but I'm realizing this might be easier to follow if the event struct declaration were moved, and the comment were changed:

```suggestion

	 /* Device control updates are a form of shadow update, so fire an
	  * NRF_CLOUD_EVT_RX_DATA_SHADOW event
	  */

	struct nrf_cloud_evt cloud_evt = {
		.type = NRF_CLOUD_EVT_RX_DATA_SHADOW
	};
	
	cloud_evt.data = evt->param.cc->data;
	cloud_evt.topic = evt->param.cc->topic;

	nfsm_set_current_state_and_notify(nfsm_get_current_state(), &cloud_evt);

	return err;
```

Reply:
In fact, should the snippet that fires `NRF_CLOUD_EVT_RX_DATA_SHADOW` be moved directly into `cc_rx_data_handler`? It always executes (except if there is an error during handling control updates), and I don't think it depends at all on the specifics of control updates